### PR TITLE
WIP: Relooper

### PR DESCRIPTION
**Pull request by simeonschaub**
An initial prototype for a relooper to address #3. The code is mostly based on the implementation in `Charlotte.jl`, but passing arguments between blocks doesn't work correctly yet. CC: @simondanisch

### CODE REVIEW

**Comment by simeonschaub** [comment]
It's now somewhat usable but this excerpt from [https://github.com/SimonDanisch/Sugar.jl/blob/master/test/runtests.jl#L6](https://github.com/SimonDanisch/Sugar.jl/blob/master/test/runtests.jl#L6) still doesn't work properly yet:
```julia
x = if b == 22
    7
else
    8
end
```

File: src/tools/relooper.jl [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+                     Expr(:block, restructure(blocks, [args[1]], symbols, entries[1], loops)...))
+               ]
+    else
+        @show entries
+        error("Rebuild error")
+    end
+end
+
+function restructure(code::IR, args=[])
+    return Expr(:block, restructure(blocks(code), [["foo", args...]])...)
```
Comment (simeonschaub) [inline src/tools/relooper.jl]: @MikeInnes Do you have any test cases that actually use the first argument passed to the entry point? It says in the docs that this should be the function itself and is used for some closures and callable objects, but I wasn't able to produce an example.

Comment (MikeInnes) [reply src/tools/relooper.jl]: You can define a closure like `f = let y = 3; x -> x*y; end` or define a struct and make it callable.

**Comment by MikeInnes** [comment]
I think it'd be nice to separate this a little from the IR data structure. I'm imagining having a `CFG` data structure that just represents control flow only, without any code, and then the relooper will produce a new CFG type (possibly an `Expr`, with structured control flow). To get back to Julia code you can then fill in the relooper-generated CFG with the content of each block.

Then we could also do things like emitting a `WASMCFG` type (or whatever) that uses WebAssembly's control flow constructs.

**Comment by simeonschaub** [comment]
That definitely sounds interesting! Unfortunately, I don't have much prior experience with codegen and am not too familiar with all the possible use cases, so I'm probably not the right person to design such a data structure, but if something like that existed, I would be willing to port this code from here. This does sound like it would take quite a bit of effort though, so for the short term it's probably easier to do it like this.

**Comment by MikeInnes** [comment]
A CFG really is just a graph. I recommend just using `CFG = Vector{Vector{Int}}`, where the vector at `i` lists the blocks which block `i` can jump to. The output can contain `Block(i)` objects rather than code.

**Comment by MikeInnes** [comment]
I had a quick go at the factored version in #25. It should be possible to port the rest of the utilities you've built here to work with the output of `reloop`, which will hopefully make them simpler.

**Comment by simeonschaub** [comment]
Sorry, I haven't had the chance to work on this lately. #25 looks really promising, I think this will definitely be the way to go. As far as I could tell, most of the functionality from this PR is actually implemented. Is there anything else left to do for #25, except for documentation and tests?

**Comment by MikeInnes** [comment]
I'd still like to clean up the output a fair bit, but basically yes.

**Comment by SimonDanisch** [comment]
Merged in: https://github.com/MikeInnes/IRTools.jl/pull/25

**Pull request closed without merging**